### PR TITLE
Vector128.ShuffleUnsafe

### LINKED ISSUE #81609: [API Proposal]: Vector128.ShuffleUnsafe
### Background and motivation

When implementing vectorized algorithms using the new `Vector128` and friends APIs, `Vector128.Shuffle` is a case where one sometimes has to drop back down to using architecture-specific instructions.

`Vector128.Shuffle` guarantees that results for out-of-range indices (values above 15) are normalized to 0.
If the JIT can't prove that indices are already in the [0, 15] range, it is forced to emit extra logic to do the normalization.
For some algorithms, you may also want exactly the behavior of the underlying intrinsic (e.g. X86's PSHUFB will treat indices MOD 16, and set the result to 0 if the high bit is set).

The solution in both cases is to use the architecture-specific instruction directly, often writing a helper such as
```csharp
private static Vector128<byte> Shuffle(Vector128<byte> vector, Vector128<byte> indices)
{
    return Ssse3.IsSupported
        ? Ssse3.Shuffle(vector, indices)
        : AdvSimd.Arm64.VectorTableLookup(vector, indices);
}
```
(as of #80963, we already have 4 such helpers in runtime)

I propose adding a `Vector128.ShuffleUnsafe` helper that does not guarantee identical behavior for out-of-range indices across platforms. Instead, it would be defined as using the underlying intrinsic directly when available.

### API Proposal

```csharp
namespace System.Runtime.Intrinsics;

public static class Vector128
{
    // Existing
    public static Vector128<byte> Shuffle(Vector128<byte> vector, Vector128<byte> indices);

    // Proposed
    public static Vector128<byte> ShuffleUnsafe(Vector128<byte> vector, Vector128<byte> indices);
}
```

### API Usage

```diff
-Vector128<byte> bitMask = Shuffle(bitmapLookup, lowNibbles);
-Vector128<byte> bitPositions = Shuffle(Vector128.Create(0x8040201008040201).AsByte(), 
+Vector128<byte> bitMask = Vector128.ShuffleUnsafe(bitmapLookup, lowNibbles);
+Vector128<byte> bitPositions = Vector128.ShuffleUnsafe(Vector128.Create(0x8040201008040201).AsByte(), highNibbles);

-private static Vector128<byte> Shuffle(Vector128<byte> vector, Vector128<byte> indices)
-{
-    return Ssse3.IsSupported
-        ? Ssse3.Shuffle(vector, indices)
-        : AdvSimd.Arm64.VectorTableLookup(vector, indices);
-}
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by AlexRadch**
Close https://github.com/dotnet/runtime/issues/81609 issue.

### CODE REVIEW

**Comment by EgorBo** [comment]
Test failures look related

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector256.cs [hunk 1]
```diff
@@ -2351,6 +2353,45 @@ public static Vector256<sbyte> Shuffle(Vector256<sbyte> vector, Vector256<sbyte>
...
+                    AdvSimd.Arm64.VectorTableLookup(table, indices.GetLower()),
+                    AdvSimd.Arm64.VectorTableLookup(table, indices.GetUpper()));
+            }
+
+            if (PackedSimd.IsSupported)
+            {
+                return Vector256.Create(
+                    PackedSimd.Shuffle(vector.GetLower(), vector.GetUpper(), indices.GetLower()),
+                    PackedSimd.Shuffle(vector.GetLower(), vector.GetUpper(), indices.GetUpper()));
+            }
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector256.cs:2390]: We shouldn't need this additional handling for `V256`/`V512`. It isn't expected that users are calling these, let alone expecting good performance, when `IsHardwareAccelerated` reports `false`.

It's just causing the JIT to spend more time doing dead code elimination, managing the inliner budget, etc.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector256.cs [hunk 2]
```diff
@@ -2351,6 +2353,45 @@ public static Vector256<sbyte> Shuffle(Vector256<sbyte> vector, Vector256<sbyte>
...
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(Avx2))]
+        [CompExactlyDependsOn(typeof(AdvSimd))]
+        [CompExactlyDependsOn(typeof(AdvSimd.Arm64))]
+        [CompExactlyDependsOn(typeof(PackedSimd))]
+        public static Vector256<byte> ShuffleUnsafe(Vector256<byte> vector, Vector256<byte> indices)
+        {
+            if (Avx2.IsSupported)
+            {
+                return Avx2.Shuffle(vector, indices);
```
Comment (MihaZupan) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector256.cs:2374]: These hardware intrinsics have different semantics than Vector256/512 Shuffles.

Avx2.Shuffle is effectively two Vector128.Shuffles on each half. The indices are still [0, 15]. Similar for Avx512.

Vector256.Shuffle on the other hand works with the whole vector and allows you to pick and index [0, 31].

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector256.cs [hunk 3]
```diff
@@ -2351,6 +2353,45 @@ public static Vector256<sbyte> Shuffle(Vector256<sbyte> vector, Vector256<sbyte>
...
 
+        /// <summary>Creates a new vector by selecting values from an input vector using a set of indices.
+        /// Behavior is platform-dependent for out-of-range indices.</summary>
+        /// <param name="vector">The input vector from which values are selected.</param>
+        /// <param name="indices">The per-element indices used to select a value from <paramref name="vector" />.</param>
+        /// <returns>A new vector containing the values from <paramref name="vector" /> selected by the given <paramref name="indices" />.</returns>
+        /// <remarks>Unlike Shuffle, this method delegates to the underlying hardware intrinsic without ensuring that <paramref name="indices"/> are normalized to [0, 31].
+        /// On hardware with <see cref="Avx2"/> support, indices are treated as modulo 32, and if the high bit is set, the result will be set to 0 for that element.
+        /// On hardware with <see cref="AdvSimd.Arm64"/> support, this method behaves the same as Shuffle.</remarks>
+        /// On hardware with <see cref="PackedSimd"/> support, indices are treated as modulo 32.
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector256.cs:2364]: nit: This remarks is confusing. `Shuffle` doesn't ensure the indices are normalized. It simply treats all out of range indices as setting the result to `0`.

`Unsafe` simply changes how `out of range` indices are handled and allows it to differ based on the target hardware. In the case of `xarch` most out of range indices are handled as mod 32. But indices with the MSB set are set to `0` instead.

We can probably exclude the Arm64/Wasm callouts for V256/V512 as they aren't important. They can be combined into a general statement that other platforms will treat this the same as `Shuffle`

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128.cs [hunk 4]
```diff
@@ -2384,7 +2384,7 @@ public static Vector128<sbyte> Shuffle(Vector128<sbyte> vector, Vector128<sbyte>
         [CompExactlyDependsOn(typeof(AdvSimd))]
         [CompExactlyDependsOn(typeof(AdvSimd.Arm64))]
         [CompExactlyDependsOn(typeof(PackedSimd))]
-        internal static Vector128<byte> ShuffleUnsafe(Vector128<byte> vector, Vector128<byte> indices)
+        public static Vector128<byte> ShuffleUnsafe(Vector128<byte> vector, Vector128<byte> indices)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128.cs:2387]: Is there anywhere else this should be consumed in dotnet/runtime as part of making it public?

Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128.cs:2387]: I believe I updated all such places when adding the internal helper as part of #80963

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128.cs:2387]: Ok.  I know you updated places in corelib.  I was wondering if there were places outside of corelib you couldn't update.  Sounds like no.

Comment (MihaZupan) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128.cs:2387]: There are two places I saw where we still use the platform-specific instructions directly ([BitArray](https://github.com/dotnet/runtime/blob/1380a0b327a24a924ef3f63f144558c742da0a3c/src/libraries/System.Collections/src/System/Collections/BitArray.cs#L844-L852) and [OptimizedInboxTextEncoder](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/OptimizedInboxTextEncoder.Ssse3.cs)), but both of those are in larger platform-specific blocks of logic, so there was no reason to replace them.

**Review by adamsitnik — CHANGES_REQUESTED** [review]
Plenty of related tests are failing. Example:

```log
Test System.Runtime.Intrinsics.Tests.Vectors.Vector256Tests.Vector256ByteShuffleOneInputWithLocalIndicesTest has failed
```

Please address the comments posted by other reviewers and fix the tests.

Thank you for your contribution!

**Pull request closed without merging**